crates/matrix-sdk/src/widget/mod.rs

```diff
@@ -50,6 +41,84 @@ pub struct Comm {
     pub to: Sender<String>,
 }
 
+/// Information about a widget.
+#[derive(Debug)]
+pub struct WidgetSettings {
+    /// Widget's unique identifier.
+    pub id: String,
+    /// Whether or not the widget should be initialized on load message
+    /// (`ContentLoad` message), or upon creation/attaching of the widget to
+    /// the SDK's state machine that drives the API.
+    pub init_on_load: bool,
+    /// This contains the url from the widget state event
+    /// In this url placeholders can be used to pass information from the client to the widget
+    /// Possible values are: `$widgetId`, `$parentUrl`, `$userId`, `$lang`, `$fontScale`, `$analyticsID`
+    /// 
+    /// # Examples
+    /// 
+    /// e.g `http://widget.domain?username=$userId`
+    /// will become: `http://widget.domain?username=@user_matrix_id:server.domain`
+    raw_url: Url,
+}
+impl WidgetSettings {
+    /// Create the actual Url that can be used to setup the WebView or IFrame that contains the widget.
+    /// 
+    /// # Arguments
+    /// 
+    /// * `room` - A matrix room which is used to query the logged in username
+    /// * `parent_url` - The parent url is used as the target for the postMessages send by the widget
+    /// Should be the url of the app hosting the widget.
+    /// In case the app hosting the widget is not a webapp the platform specific 
+    /// value needs to be used or `"*"` a wildcard.
+    /// Be aware that this means the widget will receive its own postmessage messages.
+    /// The (js) matrix-widget-api ignores those however so this works but it might break
+    /// custom implementations. So always keep this in mind.
+    /// * `font_scale` - The font scale used in the widget. 
+    /// This should be in sync with the current client app configuration
+    /// * `lang` - the language e.g. en-us
+    /// * `analytics_id` - This can be used in case the widget wants to connect to the 
+    /// same analytics provider as the client app only set this value on widgets which are known.
+    pub fn get_url(
+        &self,
+        room: JoinedRoom,        
+        parent_url: String,
+        font_scale: f64,
+        lang: String,
+        analytics_id: String,
+    ) -> String {
+        self.raw_url
+            .as_str()
+            .replace("$widgetId", &self.id)
+            .replace("$parentUrl", &encode(parent_url.as_str()))
+            .replace("$userId", &room.client().user_id().unwrap().to_string())
+            .replace("$lang", &lang)
+            .replace("$fontScale", &font_scale.to_string())
+            .replace("$analyticsID", &analytics_id)
+    }
```

Comment:
I may be wrong about it (I think @jplatte may be more aware of how things are done in the Rust SDK and provide a more useful feedback), but typically I would expect that if `lang` is limited to some fixed inputs (`en-us`, `en-uk` etc), then `lang` should be an `enum`, the `parent_url` should probably be `http::Uri`.

Also, the `user_id().unwrap()` looks quite scary here (since there is no comment that says why it would never fail - it looks like it could panic).

Reply:
Here I have the same question again. What exactly is the difference between a rust Uri and Url